verilog/analysis/checkers/disable_fork_rule.cc

```diff
@@ -0,0 +1,124 @@
+
+#include "verilog/analysis/checkers/disable_fork_rule.h"
+
+#include <set>
+#include <string>
+
+#include "absl/strings/str_cat.h"
+#include "absl/strings/string_view.h"
+#include "common/analysis/citation.h"
+#include "common/analysis/lint_rule_status.h"
+#include "common/analysis/matcher/bound_symbol_manager.h"
+#include "common/analysis/matcher/core_matchers.h"
+#include "common/analysis/matcher/matcher.h"
+#include "common/text/symbol.h"
+#include "common/text/syntax_tree_context.h"
+#include "verilog/CST/identifier.h"
+#include "verilog/CST/verilog_matchers.h"  // IWYU pragma: keep
+#include "verilog/CST/verilog_nonterminals.h"
+#include "verilog/analysis/descriptions.h"
+#include "verilog/analysis/lint_rule_registry.h"
+
+namespace verilog {
+namespace analysis {
+
+using verible::GetStyleGuideCitation;
+using verible::LintRuleStatus;
+using verible::LintViolation;
+using verible::SyntaxTreeContext;
+using verible::matcher::Matcher;
+
+// Register DisableForkNoLabelsRule
+VERILOG_REGISTER_LINT_RULE(DisableForkNoLabelsRule);
+
+absl::string_view DisableForkNoLabelsRule::Name() {
+  return "disable-statement";
+}
+const char DisableForkNoLabelsRule::kTopic[] =
+    "disable-labelled-fork-statement";
+const char DisableForkNoLabelsRule::kMessage[] =
+    "Invalid usage of disable statement. Allowed construction is: disable "
+    "fork;";
+
+std::string DisableForkNoLabelsRule::GetDescription(
+    DescriptionType description_type) {
+  return absl::StrCat(
+      "Checks that there are no occurrences of ",
+      Codify("disable some_label", description_type),
+      " if label is referring to a fork or other none sequential block label.",
+      ". Use ", Codify("disable fork", description_type), " instead. See ",
+      GetStyleGuideCitation(kTopic), ".");
+}
+
+static const Matcher& DisableMatcher() {
+  static const Matcher matcher(
+      NodekDisableStatement(DisableStatementHasLabel()));
+  return matcher;
+}
+
+void DisableForkNoLabelsRule::HandleSymbol(const verible::Symbol& symbol,
+                                           const SyntaxTreeContext& context) {
+  verible::matcher::BoundSymbolManager manager;
+  if (DisableMatcher().Matches(symbol, &manager)) {
+    // if no kDisable label, return, nothing to be checked
+    const auto disableLabels = FindAllSymbolIdentifierLeafs(symbol);
+    if (disableLabels.size() == 0) {
+      return;
+    }
+    // look for every kBegin node starting from kDisableLabel token
+    // the kDisableLabel can be nested in some kBegin nodes
+    // so we're looking for the kBegin node that direct parent
+    // is not one of the initial/final/always statemnts since such blocks are
+    // considered to be invalid. If the label for diable statment is not
+    // found, it means that there is no appropriate label or the label
+    // points to the illegal node such as forked label
+    const auto& rcontext = reversed_view(context);
+    for (auto rc = rcontext.begin(); rc != rcontext.end(); rc++) {
```

Comment:
I like the new structure with 'early out' continues. It makes the code easy to follow.